src/neo/Network/P2P/Payloads/Header.cs

```diff
@@ -1,16 +1,81 @@
+using Neo.IO;
+using Neo.IO.Json;
+using Neo.Ledger;
+using Neo.Persistence;
+using Neo.SmartContract;
+using Neo.SmartContract.Native;
+using Neo.Wallets;
 using System;
 using System.IO;
 
 namespace Neo.Network.P2P.Payloads
 {
-    public class Header : BlockBase, IEquatable<Header>
+    public sealed class Header : IEquatable<Header>, IVerifiable
     {
-        public override int Size => base.Size + 1;
+        public uint Version;
+        public UInt256 PrevHash;
+        public UInt256 MerkleRoot;
+        public ulong Timestamp;
+        public uint Index;
+        public byte PrimaryIndex;
+        public UInt160 NextConsensus;
+        public Witness Witness;
 
-        public override void Deserialize(BinaryReader reader)
+        private UInt256 _hash = null;
+        public UInt256 Hash
         {
-            base.Deserialize(reader);
-            if (reader.ReadByte() != 0) throw new FormatException();
+            get
+            {
+                if (_hash == null)
+                {
+                    _hash = this.CalculateHash();
+                }
+                return _hash;
+            }
+        }
+
+        public int Size =>
+            sizeof(uint) +      // Version
+            UInt256.Length +    // PrevHash
+            UInt256.Length +    // MerkleRoot
+            sizeof(ulong) +     // Timestamp
+            sizeof(uint) +      // Index
+            sizeof(byte) +      // PrimaryIndex
+            UInt160.Length +    // NextConsensus
+            1 + Witness.Size;   // Witness   
+
+        Witness[] IVerifiable.Witnesses
+        {
+            get
+            {
+                return new[] { Witness };
+            }
+            set
+            {
+                if (value.Length != 1) throw new ArgumentException();
+                Witness = value[0];
+            }
+        }
+
+        public void Deserialize(BinaryReader reader)
+        {
+            ((IVerifiable)this).DeserializeUnsigned(reader);
+            Witness[] witnesses = reader.ReadSerializableArray<Witness>(1);
+            if (witnesses.Length != 1) throw new FormatException();
+            Witness = witnesses[0];
+        }
+
+        void IVerifiable.DeserializeUnsigned(BinaryReader reader)
+        {
+            Version = reader.ReadUInt32();
```

Comment:
We can use only a byte for version or at least check it

```suggestion
            Version = reader.ReadUInt32();
            if (Version > 0) throw new FormatException();
```